feat: add Spring and Tween classes

Rich-Harris opened the pull request:
This adds a `Spring` class as an alternative to the existing `spring` store factory. The behaviour is essentially identical, but the API is slightly different:

```diff
-const coords = spring({ x: 50, y: 50 }, opts);
+const coords = new Spring({ x: 50, y: 50 }, opts);

// set new value immediately
-coords.set(value, { hard: true });
+coords.set(value, { instant: true });

// set new value but preserve momentum for 500 milliseconds (useful for 'throwing' interactions)
-coords.set(value, { soft: true });
+coords.set(value, { preserveMomentum: 500 });
```

`instant` and `preserveMomentum` are clearer than `hard` and `soft`, which seem like they're somehow symmetrical but are in fact unrelated.

To keep a spring in sync with some other value (such as a prop), use `Spring.of`:

```diff
let { number } = $props();

-const thing = spring();
-$: thing.set(number);
+const thing = Spring.of(() => number);
```

In this case it's still possible to do `thing.set(value, opts)`, since you might (e.g.) need to use `{ instant: true }`, and making it 'readonly' in the function case could be overly restrictive.

`spring.set` will return a promise that resolves when `spring.current` reaches `spring.target`. If you don't need the promise, and don't need to pass `instant` or `preserveMomentum` options, you can manipulate `spring.target` directly. (This does mean we have two ways of doing things, but we need to expose `spring.target` anyway and it would be weird if it was readonly.)

A nice thing about using classes: you can do `const spring = new Spring(...)` instead of always having to come up with a more descriptive name.

TODO

- [x] add a `Tween` class to go with `Spring`
- [x] docs
- [ ] tests? (though there are very few `spring` tests to adapt — not totally sure what tests would look like here)

### Before submitting the PR, please make sure you do the following

- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] Prefix your PR title with `feat:`, `fix:`, `chore:`, or `docs:`.
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests and linting

- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint`

abdel-17 commented:
> As an alternative to providing an initial value, you can provide a function, making it easy to keep a spring in sync with some other value:

Correct me if I'm wrong, but the value of the spring would get out of sync with the external state if you try to set it directly, no?
```ts
let { number } = $props();

const thing = new Spring(() => number);

thing.set(5); // oops now we’re out of sync
```

I think a value change callback could be added to avoid such cases.
```ts
let { number = $bindable() } = $props();

const thing = new Spring(() => number, {
  onChange: (v) => (number = v)
});

thing.set(5); // still in sync
```

Rich-Harris commented:
> the value of the spring would get out of sync with the external state if you try to set it directly, no?

Yes, and the solution is 'don't do that'. In general you wouldn't need to `set` a store with a function input, but it's a useful escape hatch if you need to do something like [this](https://svelte-5-preview-git-new-spring-svelte.vercel.app/#H4sIAAAAAAAAE11SX2vDIBD_KocMkrKyZK9p7Nhn2OO6h9RehtSo6KWjSL77NKbJGMHI6e_fHQbWS4WeNZ-B6W5A1rB3a9me0d2mwt9QEcbam9GJdNJ64aSl40mfSA7WOIIAH9ZJ_Q0T9M4MUGRWNRiSRheHBE1LIYF15tuh98DhyVNHWNa7w-PSZxkOGn8WzbLcAT-utBkb0f2oRRIHqYXDATVFXJiNSPZQbjacw2u93p0oe7x4pLLex-hSxxiaGiA3Ikw5TPr-JK2XwwlQedy0VshzNHlg0hZ_bbUNSrfnkSimNVooKa48rLGneZDhX154gyKWSAU0UMwcGLCYZbNUll1pt06NyMPSnBidi9pQRanp2FYPWCZ5uiucbVf60tFFequ6ewNnZcT1sHUyM-IrGMxF9hIvrEnTmr6mXxrCSUVAAgAA):

```js
let progress = $state(0);
let spring = new Spring(() => progress);

function increment() {
  if (progress === 10) {
    spring.set(0, { instant: true });
    progress = 0;
  } else {
    progress += 1;
  }
}
```

It would be strange to have an `onChange` callback that updated to the _target_ value rather than the _current_ value.

abdel-17 commented:
Yeah fair. I just worry this might cause weird bugs because you expect the values to always be in sync.

PuruVJ commented:
Possible to sneak in this as well? https://github.com/sveltejs/svelte/issues/9141#issuecomment-2080643588

jeremy-deutsch commented:
Because of the function case, it feels like it might be better to have some kind of `cancelAnimation()`/`jump()` method instead of the `{ immediate: true }` option?

Like:
```
<script>
  let value = $state(0);

  const spring = new Spring(() => value);

  function resetSpring() {
    value = 0;
    // switch directly to the current value without animating
    spring.cancelAnimation();
  }
</script>
```

This works even if you're using `set()`:
```
<script>
  const spring = new Spring(0);

  function resetSpring() {
    spring.set(0);
    // switch directly to the current value without animating
    spring.cancelAnimation();
  }
</script>
```

I like that this way you no longer have to know what the final spring position is in order to skip the animation. That seems really nice for the function case.

I also think a `preserveMomentum(seconds)` method could work similarly.

ottomated commented:
I'd love to see this in 5.0 and would like to work on it, @Rich-Harris do you have any input on the proposed APIs or should I just work on merge conflicts and the Tween class?

newsve commented:
In this context: My app heavily relies on many sophisticated spring animations and Sveltte 4 custom stores (based on writable, etc.). I actually created kind of an rendering engine based on Svelte's building blocks.

I believe that this new spring class adapts all subtleties from the previous API/implementation, e.g. hard and soft props, stiffness and dampening, still I am a bit afraid that I can't replicate all the motions, also there would be some significant refactoring effort, basically rewriting the whole thing. I am still refactoring hundreds of legacy .svelte files to Svelte 5. BTW, Svelte 5 was really drop-in for me and the legacy mode is a excellent way to nudge the community to a migration without breaking changes! 🙂

@Rich-Harris do you already know if you gonna remove (1) the prior spring API and (2) and the Svelte 4 stores, writable etc. in Svelte 6 or in any version after 6? Just to be prepared to plan our future, looking fwd to your reply!

Rich-Harris commented:
A few changes:

- Instead of `new Spring(value)` and `new Spring(fn)`, it's now `new Spring(value)` and `Spring.of(fn)`. I think the clearer separation is better — we can now say that you can create a `new Spring(...)` anywhere (including in a shared module) but can only call `Spring.of(...)` inside an effect root (i.e. during component initialisation)
- `spring.target` is exposed. It would be very inconvenient if you always had to go through `spring.set(...)`. This does break the 'only have one way of doing things' rule but I think it's worth it
- `preserveMomentum` is now in milliseconds
- I added a `Tween` class that behaves the same way

One problem: `svelte/motion` already exports `Spring` and `Tweened` interfaces. I can't think of a better name for the `Spring` class and don't think we should add another module, so in this PR I've renamed the existing interfaces to `SpringStore` and `TweenedStore` to avoid confusion. This is a breaking change, and there _are_ a handful of [uses in the wild](https://grep.app/search?q=Spring&case=true&words=true&filter[lang.pattern][0]=svelte&filter[lang][0]=Svelte), so I won't be surprised if someone vetoes this. I would be eager to hear better ideas though.

Also, this PR deprecates the existing `spring` and `tweened` stores. I think that's the right move, but if people feel strongly we can delay this. @newsve to answer your question, these stores will continue to exist for a good while yet — they will _eventually_ be removed, but not until Svelte 5 has been out for at least a year or so.

- [updated docs](https://svelte-dev-git-preview-svelte-11519-svelte.vercel.app/docs/svelte/svelte-motion)
- [`Spring` demo](https://svelte.dev/playground/hello-world?version=pr-11519#H4sIAAAAAAAAE5WPMWvDMBBG_8pxSxIocWbXDnTL3rHuIFsXI5BP4nROUoz-e-U60KVL14_3HncLspkIa7yQ9wHuQbyFPVmnZA_4glfnKWH9saB-xZVbh7I_rbcYj-lGXtetN4n-2ofASqwlg00axEU9d9ypm2IQhQXeozgeIcNVwgS7zaumoC7w7rWgBS6NpJA2sgWm-1Pbnw4FaarfMDeO46ywHtx2KIZH6hB6x7a-GT9Tu2ydoxoZSTNU539IwyxSvslgXTK9J_ujlzeVHoq1ykz5M38DYL5OlVgBAAA=)
- [`Spring.of` demo](https://svelte.dev/playground/hello-world?version=pr-11519#H4sIAAAAAAAAE42QwWqFMBBFf2UYCk-haNdWhbfrvsvaRTTjIxAnIRlfW8R_b1KFbrro9uace5lsyGohbPCFrHXw4YLVUJA2QrrER5yNpYjN24by5TOXg5Sf1tX7Kt7JSs5GFemvfHIsxJJqsI1TMF76gQcxi3dBYINXHwzfYIc5uAUuh1cvTozjy3NCE2xJgNdlpAAdPERRQsVTeT6mgSgQj5ru7KvcXBQldP3pZbitf_e5NexXgXxXN2BQfKMBYTSsm7uyK3XbIe5Q9_-hj_1qWkNI1-6gTVSjJf2jp28Q-hRsJKy0v-_fiyIspHgBAAA=)

abdel-17 commented:
I think the function case should expose an `onChange` option to avoid the values from getting out of sync.
```ts
const spring = Spring.of(() => value, {
  onChange(v) {
    value = v;
  }
});
```

This would get called when the spring's value reaches its target to avoid calling the callback repeatedly.

Rich-Harris commented:
What for? You can just track `spring.current` in an effect
